python/pyarrow/fs.py

```diff
@@ -356,7 +356,12 @@ def get_file_info_selector(self, selector):
             selector.base_dir, maxdepth=maxdepth, withdirs=True, detail=True
         )
         for path, info in selected_files.items():
-            infos.append(self._create_file_info(path, info))
+            # Need to exclude base directory from selected files if present
+            # (fsspec filesystems, see GH-37555)
+            if (path != selector.base_dir and
+                path != "/" + selector.base_dir[:len(selector.base_dir)-1] and
+                    path != selector.base_dir[:len(selector.base_dir)-1]):
```

Comment:
Is this logic to handle starting/ending slashes `/`? If so, it might be easier to read with start/end vars e.g.
```
start = 0
end = len(selector.base_dir)
if selector.base_dir.startswith("/"):
    start += 1
if selector.base_dir.endswith("/"):
    end -= 1
if path != selector.base_dir[start:end]:
    ...
```

Reply:
Yes, correct! Great, will change 👍

Reply:
Unfortunately starting/ending slashes also change in the `path` - I have forgot about that.

- for `fsspec.LocalFileSystem()` the `path` and `base_dir` are equal: '/.../test_get_file_info_with_select0/selector-dir'
- for `fsspec.filesystem("memory")` the `path` and `base_dir` differ like so: '/selector-dir' vs 'selector-dir/'
- for `s3fs.S3FileSystem()` the `path` and `base_dir` differ like so: 'pyarrow-filesystem/selector-dir' vs 'pyarrow-filesystem/selector-dir/'

I could read with start/end for both paths, if you still feel it will be clearer?

```python
```

Reply:
Starting/ending slashes also change in the path - I have forgot about that.

- for `fsspec.LocalFileSystem()` the `path` and `base_dir` are equal: '/.../test_get_file_info_with_select0/selector-dir'
- for `fsspec.filesystem("memory")` the `path` and `base_dir` differ like so: '/selector-dir' vs 'selector-dir/'
- for `s3fs.S3FileSystem()` the `path` and `base_dir` differ like so: 'pyarrow-filesystem/selector-dir' vs 'pyarrow-filesystem/selector-dir/'

Used start for both paths, hope it is clear.